server/controllers/v1/post.ts

```diff
@@ -0,0 +1,195 @@
+      // in a list, so ignoring body is ok as it won't be displayed anyway.
+      attributes: ['id', 'title', 'feedbackScore', 'createdAt'],
+      order: [['createdAt', 'DESC']],
+      offset: offset,
+    });
+
+    return {status: data.count > 0 ? 200 : 204, data: {result: data.rows, count: data.count}};
+  }
+
+  /**
+   * Search for the post by the given ID if it exists.
+   *
+   * @param postID - The identifier used to find the specific post.
+   * @returns The details of the post
+   */
+  async getPost(
+    postID: string
+  ): Promise<{status: number; data: {result?: Post; message?: string}}> {
+    const data = await this.postsRepo.findByPk(postID);
+
+    if (!data) {
+      return {
+        status: 404,
+        data: {message: `Post ${postID} could not be found`},
+      };
+    }
+    return {
+      status: 200,
+      data: {result: data},
+    };
+  }
+
+  /**
+   * Delete the post by a given ID.
+   *
+   * @param postID - The identifier of the post to destroy.
+   * @returns A status object indicating the results of the action.
+   */
+  async deletePost(
+    userId: string,
+    postID: string
+  ): Promise<{status: number; data?: {message?: string}}> {
+    const result = await this.postsRepo.findOne({where: {id: postID}});
+
+    if (!result) {
+      return {status: 404, data: {message: `Post ${postID} could not be deleted.`}};
+    } else if (result.UserId != userId) {
+      return {status: 401, data: {message: 'Unauthorized to delete the post.'}};
+    }
+    await result.destroy();
+    return {status: 204};
+  }
+
+  /**
+   * Change the vote of a given post
+   * @param postID - The identifier of the post the vote on.
+   * @param amount - 1 if it should be incremented, -1 if it should be decremented.
+   * @returns A status object indicating if the action was successful.
+   */
+  private async vote(
+    postID: string,
+    amount: 1 | -1
+  ): Promise<{status: number; data?: {result?: Post; message?: string}}> {
+    // TODO: We should track which users voted and how frequently, and take action to prevent vote
+    // spamming.
+    const data = await this.getPost(postID);
+
+    if (data.status != 200) {
+      return data;
+    }
+
+    try {
+      (data.data.result!.feedbackScore as number) += amount;
+      await data.data.result!.save();
+      return {status: 204, data: {result: data.data.result}};
+    } catch (err) {
+      console.error(`Could not change vote for: ${postID}`);
+      return {
+        status: 400,
+        data: {message: `Could not ${amount == 1 ? 'upvote' : 'report'} post: ${postID}`},
+      };
+    }
+  }
+
+  /**
+   * Downvote a given post.
+   *
+   * @param postID - The post to downvote.
+   * @returns A status object indicating whether the post was downvoted.
+   */
+  async report(
+    postID: string
+  ): Promise<{status: number; data?: {result?: Post; message?: string}}> {
+    return this.vote(postID, -1);
```

Comment:
Are we treating reports as simply downvotes?

If we're going to allow downvotes, I think we should either have some system in place to automatically remove a post with too many downvotes, or we should have downvoting and a separate report functionality that might be considered more seriously than a downvote. It would probably also require user authentication first. What do you guys think?

Reply:
We can just have "like" functionality, and "report'. People can like/unlike a post and the count is visible to the public. If a post is reported, we keep track of that data in the backend (and show nothing to the frontend). A post with too many reports is deleted.

Reply:
Wouldn't report count be another field in the model?

Reply:
Yea you're right, nevermind. We can keep this, reports will downvote a post and likes will increase the score of the post. We can set a threshold where a post gets removed after being too low

Reply:
To prevent vote spamming we also need to keep track of which users voted on what. So rather than being a score, we need another model either way. I think keeping this as-is for now makes sense, then in a future PR introducing a separate association between user and vote (userid, postid would be the two fields).

Reply:
Yes, I agree. The model can also include interactions with 'checking in' so we can prevent them from checking into an event multiple times.

Reply:
Ok. I'm merging this in so we can start connecting things up.